AeroDyn driver update for multiple wind turbines, with arbitrary motions and geometries

ebranlard opened the pull request:
This pull request is almost ready to merge. 


**Feature or improvement description**
New AeroDyn driver functionalities:
- Multiple turbines with arbitrary number of blades and orientation (advanced rotor inputs)
- Different motions for the base, nacelle yaw, and individual blade pitch: fixed values, sinusoidal, or arbitrary via time series 
- Simple wind definition (Power law), or InflowWind 


**Old features still supported:**
- Basic rotor inputs for HAWT (simply provide HubHeight, Tilt, Tower2Shaft)
- Prescribed Time series of Wind Speed / PL Exp / Pitch / Yaw  (`AnalysisType=2`)
- Combined case analyses (`AnalysisType=3`)


**Example of Applications**
- Sinusoidal motions of the Platform base
- Vertical axis turbines
- Multiple wind turbines
- Quad rotor
- Kites

![110567794-4c540780-810f-11eb-9cf2-23273877ec0f](https://user-images.githubusercontent.com/1318316/111832807-db52e380-88b6-11eb-859c-95b51283370b.gif)



**Impacted areas of the software**
AeroDyn, OLAF

**Additional supporting information**

- [Live documentation](https://ebranlard-openfast.readthedocs.io/en/f-driver/source/user/aerodyn/driver.html)  (will be updated soon)
-  [Test/examples files](https://github.com/OpenFAST/r-test/tree/f/driver/modules/aerodyn) files for all the different features with README describing each cases.


**Check list**

- [x] Merge #672 
- [x] Merge #589
- [x] Merge #648
- [x] Finalize documentation
- [x] Merge/update new r-tests
- [x] Review
Extra features:
- [x] WrVTK harmony with OpenFAST
- [x] InitRot before debug outputs

**Test results, if applicable**
Expected some small differences for OLAF Elliptical wing, will be resolved after #648 is merged

jjonkman commented on docs/source/user/aerodyn/driver.rst:
```diff
@@ -0,0 +1,497 @@
+    1   NumTurbines  - Number of turbines (should be 1 for AnalysisType=2 or AnalysisType=3)
+
+As noted in the comment, the number of turbine should be 1 for `AnalysisType=2` and `AnalysisType=3`.
+After this section, the geometry and motion is provided for each turbine. Inputs for each turbine must have the suffix `(i)` where `i` is the turbine number.
```
Add "(even with NumTurbines = 1, then i=1)"

jjonkman commented on docs/source/user/aerodyn/driver.rst:
```diff
@@ -0,0 +1,497 @@
+
+An example of configuration with two wind turbines is shown in :numref:`fig:MultiRotor`. The figure defines the different frames and origin associated with each turbine: the turbine base frame (t), nacelle frame (n), hub frame (h), and blade frames (b). The notations and conventions follow the OpenFAST frames, except that the turbine frame does not have its origin at the tower base.
+Prescribed motions of the turbine occur at the turbine origin.
+Yawing occurs around the :math:`z_n` axis,  the blade rotates about the :math:`x_h` axis, and pitching occurs around the individual :math:`z_b` axes. The definitions of the different frames are standardized when using a basic (HAWT) input format definition, and are arbitrarily defined when using the advanced input format. More details are given in the next paragraph.
```
Change "blade rotates" to "rotor rotates" and add "blade" before pitching.

jjonkman commented on docs/source/user/aerodyn/driver.rst:
```diff
@@ -0,0 +1,497 @@
+
+Two turbine input formats are supported:
+
+- basic (`BasicHAWTFormat=True`): Basic horizontal axis wind turbine (HAWT) format. In this format, the turbine geometry is entirely determined by the number of blades (`NumBlades`), the hub radius (`HubRad`), the hub height (`HubHt`), the overhang (`Overhang`), the shaft tilt (`ShftTilt`) and the precone (`Precone`). The definition of each parameter follows the OpenFAST convention. Additionally, ...
```
I would say OpenFAST does not have a convention; perhaps you mean "ElastoDyn convention"?

jjonkman commented on docs/source/user/aerodyn/driver.rst:
```diff
@@ -0,0 +1,497 @@
+
+- advanced (`BasicHAWTFormat=False`): The position and orientation of the tower base, nacelle, hub, and individual blades can be arbitrarily defined. This can be used for HAWT and any other turbine concepts. 
+  The definition of the different frames are given in :numref:`fig:MultiRotor`.
+  The position (`BaseOriginInit`) and orientation (`BaseOriginInit`) of the turbine base frame are defined with respect to the global frame. Orientations are given using the values of three successive rotations (x-y-z Euler angles). If the base undergoes a motion, the orientation of the base frame will consists of the time-varying rotations followed by these initial rotations.
```
Change "angles" to "angle sequence".
Change "consists" to "consist".

jjonkman commented on docs/source/user/aerodyn/driver.rst:
```diff
@@ -0,0 +1,497 @@
+  The position (`BaseOriginInit`) and orientation (`BaseOriginInit`) of the turbine base frame are defined with respect to the global frame. Orientations are given using the values of three successive rotations (x-y-z Euler angles). If the base undergoes a motion, the orientation of the base frame will consists of the time-varying rotations followed by these initial rotations.
+
+  A flag indicating whether the turbine has a tower is given on the next line (`HasTower`). This flag currently affects the VTK outputs and does not have an impact on AeroDyn yet. The user still has to provide tower inputs data in AeroDyn for each turbine (see :numref:`ad_inputs_multirot`).
+  The next line indicates which projection AeroDyn is to use in its calculation. It is recommended to use `HAWTProjection=True` for HAWT, which is the default projection used in OpenFAST (projects on the coned-pitched axis). For other rotor concepts, set `HAWTprojection=False`.
```
Add "BEMT" before "calculation".
Change "OpenFAST" to "AeroDyn".

ebranlard replied:
Some of the outputs of AeroDyn rely on this projection convention (BEMT or FVW).

jjonkman commented on docs/source/user/aerodyn/driver.rst:
```diff
@@ -0,0 +1,497 @@
+
+  A flag indicating whether the turbine has a tower is given on the next line (`HasTower`). This flag currently affects the VTK outputs and does not have an impact on AeroDyn yet. The user still has to provide tower inputs data in AeroDyn for each turbine (see :numref:`ad_inputs_multirot`).
+  The next line indicates which projection AeroDyn is to use in its calculation. It is recommended to use `HAWTProjection=True` for HAWT, which is the default projection used in OpenFAST (projects on the coned-pitched axis). For other rotor concepts, set `HAWTprojection=False`.
+  The following lines indicates the position and orientations of the tower, nacelle and hub. 
```
Change "indicates" to indicate".

jjonkman commented on docs/source/user/aerodyn/driver.rst:
```diff
@@ -0,0 +1,497 @@
+  The tower stations defined in the AeroDyn input file are assumed to be given with respect to the tower origin, unlike OpenFAST which uses ground/MSL as a reference (see :numref:`ad_inputs_multirot`).
+  The hub is defined with respect to the nacelle origin and frame (n).
+
+  The definitions of the blades follows, starting with the number of blades `NumBlades`. A rotor with zero blade is supported and can be used to model an isolated tower.
```
Change "follows" to follow".
Change "blade" to "blades".

jjonkman commented on docs/source/user/aerodyn/driver.rst:
```diff
@@ -0,0 +1,497 @@
+  The hub is defined with respect to the nacelle origin and frame (n).
+
+  The definitions of the blades follows, starting with the number of blades `NumBlades`. A rotor with zero blade is supported and can be used to model an isolated tower.
+  If tower shadow/potential is used in AeroDyn, then the isolated tower will disturbed the flow of the vortex wake when OLAF is used.
```
Change "disturbed" to "disturb".

jjonkman commented on docs/source/user/aerodyn/driver.rst:
```diff
@@ -0,0 +1,497 @@
+
+  The definitions of the blades follows, starting with the number of blades `NumBlades`. A rotor with zero blade is supported and can be used to model an isolated tower.
+  If tower shadow/potential is used in AeroDyn, then the isolated tower will disturbed the flow of the vortex wake when OLAF is used.
+  When BEM is used, the flow of the blades of a given turbine are disturbed only by the turbine's tower.
```
Change "the turbine's" to "that turbine's"

jjonkman commented on docs/source/user/aerodyn/driver.rst:
```diff
@@ -0,0 +1,497 @@
+  If tower shadow/potential is used in AeroDyn, then the isolated tower will disturbed the flow of the vortex wake when OLAF is used.
+  When BEM is used, the flow of the blades of a given turbine are disturbed only by the turbine's tower.
+  The inputs for turbine `i` and blade `j` are labelled `(i_j)`.
+  The origin `BldOrigin_h`) and orientation (`BldOrientation_h`) of each blade are given with respect to the hub origin and frame (h).
```
Add "(" before BldOrigin_h.

jjonkman commented on docs/source/user/aerodyn/driver.rst:
```diff
@@ -0,0 +1,497 @@
+The turbine base motion is applied at each time step before applying the initial position and orientation of the turbine base.
+A sinusoidal motion implies that one degree of freedom (`DegreeOfFreedom`) of the turbine base is moving according to a sine function of a given amplitude (`Amplitude`) and frequency (`Frequency`), with zero phase.
+The 6 possible degrees of freedom correspond to translations or rotations of the base frame in global coordinates (g) (e.g. surge, sway, heave, roll, pitch, yaw).
+An arbitrary motion is specified via a CSV file (`BaseMotionFileName`) which contains 19 columns: time, 3 translations (global), three successive rotations (global), 3 translation velocities, 3 rotational velocities (omega, in global), 3 translational accelerations and 3 rotational accelerations (alpha, in global). Example of arbitrary input files are given in :numref:`ad_inputfiles_examples`.
```
Add "The displacements/orientations, velocities, and accelerations are not checked internally for consistency."

jjonkman commented on docs/source/user/aerodyn/driver.rst:
```diff
@@ -0,0 +1,497 @@
+The base motion is given in the same way for basic or advanced geometries.
+The motion of the base may be: fixed (`BaseMotionType=0`), sinusoidal (`BaseMotionType=1`) or arbitrary (`BaseMotionType=2`). 
+The turbine base motion is applied at each time step before applying the initial position and orientation of the turbine base.
+A sinusoidal motion implies that one degree of freedom (`DegreeOfFreedom`) of the turbine base is moving according to a sine function of a given amplitude (`Amplitude`) and frequency (`Frequency`), with zero phase.
```
Is "Frequency" in rad/s or Hz?

jjonkman commented on docs/source/user/aerodyn/driver.rst:
```diff
@@ -0,0 +1,497 @@
+    1         BldPitch(1)      - Blades pitch (deg)
+
+- advanced: When an advanced geometry is provided and when the number of blades is non-zero, the motion section contains options for the nacelle motion, rotor motion and individual blade pitch motion. 
+  The syntax for each of these motion consists of defining a type (fixed or time-varying), a value for the fixed case or a file for the time-varying case.
```
Change "motion" to "motions".

jjonkman commented on docs/source/user/aerodyn/driver.rst:
```diff
@@ -0,0 +1,497 @@
+- advanced: When an advanced geometry is provided and when the number of blades is non-zero, the motion section contains options for the nacelle motion, rotor motion and individual blade pitch motion. 
+  The syntax for each of these motion consists of defining a type (fixed or time-varying), a value for the fixed case or a file for the time-varying case.
+  The input files are CSV files containing time, position, speed and acceleration. Examples of files are given in :numref:`ad_inputfiles_examples`.
+  If the pitch rate, yaw rate, and accelerations are not known, the user should input zeros in these columns.
```
I'm not sure this needs to be stated (may cause confusion).  Instead, I would say--as above--"The displacements/orientations, velocities, and accelerations are not checked internally for consistency."

jjonkman commented on docs/source/user/aerodyn/driver.rst:
```diff
@@ -0,0 +1,497 @@
+  The syntax for each of these motion consists of defining a type (fixed or time-varying), a value for the fixed case or a file for the time-varying case.
+  The input files are CSV files containing time, position, speed and acceleration. Examples of files are given in :numref:`ad_inputfiles_examples`.
+  If the pitch rate, yaw rate, and accelerations are not known, the user should input zeros in these columns.
+  The time vector in the motion file has to be ascending, but does not need to be linear. Linear interpolation is used by the driver to determine inputs at a given time.
```
These comments should added to the base motion section as well.

jjonkman commented on docs/source/user/aerodyn/driver.rst:
```diff
@@ -0,0 +1,497 @@
+
+**Time-dependent analysis**
+
+Time-dependent analyses are used to vary few standard variables during the simulation.
```
Add "a" before "few".

jjonkman commented on docs/source/user/aerodyn/driver.rst:
```diff
@@ -0,0 +1,497 @@
+    ----- Output Settings -------------------------------------------------------------------
+      "ES15.8E2"     OutFmt      - Format used for text tabular output, excluding the time channel.  Resulting field should be 10 characters. (quoted string)
+    2                OutFileFmt  - Format for tabular (time-marching) output file (switch) {1: text file [<RootName>.out], 2: binary file [<RootName>.outb], 3: both}
+    0                WrVTK       - VTK visualization data output: (switch) {0=none; 1=animation}
```
Perhaps this doesn't matter, but OpenFAST uses WrVTK = 1 for initialization data only and 2 for animation

ebranlard replied:
changed to match openfast convention now

jjonkman commented on docs/source/user/aerodyn/driver.rst:
```diff
@@ -0,0 +1,497 @@
+
+The legacy AeroDyn format requires a minimum of three blade file names. 
+For this reason, the blades of all rotors are currently indicated successively in the `ADBlFile` list. 
+The list is populated by looping on turbines and turbine blades, with tine blades being the fast index.
```
What is "tine"?  Suggest rewording.

ebranlard replied:
I'm not sure what happened there.. The blade index is the fastest index.

jjonkman reviewed (commented):
Thanks @ebranlard.  Overall, this looks like a powerful extension to the AeroDyn driver.  I've spent some time reviewing the updated documentation.  Please find my edits and comments here.

ebranlard commented on docs/source/user/aerodyn/driver.rst:
```diff
@@ -0,0 +1,497 @@
+
+The inflow can be provided in two ways:
+
+- basic (`CompInflow=0`): uniform wind with a power law shear. The wind is defined using a reference height (`RefHt`), a powerlaw exponent (`PLExp`), and the wind speed at reference height (`HWindSpeed`). With `AnalysisType=1`, the reference wind speed and power law are defined separately as time series (see "time-dependent analysis"). With `AnalysisType=3`, these parameters are provided in a sep ...
```
Yes, it was a typo, with AnalysisType=2 or 3, PLExp and HWindSpeed are not used.

andrew-platt commented on modules/aerodyn/src/AeroDyn.f90:
```diff
@@ -2780,145 +2806,76 @@ SUBROUTINE Init_FVWmodule( InputFileData, u_AD, u, p, x, xd, z, OtherState, m, E
+   integer(intKi)                               :: iR, iB, iW, iW_incr  ! rotor index, blade index, wing index
+   integer(IntKi)                               :: ErrStat2
+   character(ErrMsgLen)                         :: ErrMsg2
+   character(*), parameter                      :: RoutineName = 'Init_FVWmodule'
```
Change routinename to `Init_OLAF`

rafmudaf commented on modules/aerodyn/src/AeroDyn_IO.f90:
```diff
@@ -1786,19 +1878,19 @@ subroutine Calc_WriteOutput_BEMT
-!         m%AllOuts( BAeroMy(k) ) = tmp(2)
-!         m%AllOuts( BAeroMz(k) ) = tmp(3)
-!      end do  ! k=blades
+      do k=1,min(p%NumBlades,4)
```
@ebranlard Why the loop to a minimum of 4 here? In other areas, the minimum is 3. Also, using `MaxBl` allows to change it all at once and prevent any inconsistencies.

rafmudaf replied:
@andrew-platt added some clarity in that these outputs are specific to VAWTS. Without context, the difference in 3 and 4 is unclear, so a comment noting this would be helpful.

ebranlard replied:
The limit is purely arbitrary for now, and not limited to VAWT. I needed more outputs (actually I needed 9 blades), so I hardcoded outputs for an extra blade (without using the matlab script). Hannah's branch will have a better handling of the outputs, this is just temporary.

rafmudaf commented on modules/aerodyn/src/AeroDyn_IO.f90:
```diff
@@ -1926,6 +2042,21 @@ subroutine Calc_WriteOutput_FVW
       end if
 
+      ! Integrate force/moments over blades by performing mesh transfer to blade root points:
+      do k=1,min(p%NumBlades,4)
```
Same here, see above

rafmudaf commented:
@ebranlard you resolve the merge conflict with the inflow wind file, I'll deal with r-test.

andrew-platt commented:
I looked at the conflict and couldn't figure out why git is complaining about it.

ebranlard commented:
It seems it was just a couple of spaces on one line. I'm not sure if this branch is pointing to the right r-tests after this merge though..

rafmudaf commented:
I think we still need to merge the r-test branch used here into that repository's `dev` branch so that we don't undo the update to r-test in https://github.com/OpenFAST/openfast/commit/792f37ee660d507191ae0807038bb791a4430c3b. Is that right?

ebranlard commented:
I've updated the r-tests, merged with the latest r-test-dev and updated the vertical axis testcase. 
I believe this pull request is ready to be merged. Maybe the r-test can be squashed?

andrew-platt reviewed (changes requested):
I think it all looks good, except the detail in `IfW_FFWind_Base.f90`

ebranlard commented on modules/inflowwind/src/IfW_FFWind_Base.f90:
```diff
@@ -485,6 +485,10 @@ SUBROUTINE GetInterpValues()
             ENDIF
          ENDIF
+         if (IYLO<0) then
+            STOP
```
I probably introduced this while debugging the error handling that was missing above. I can remove it now.

ebranlard reviewed (commented):
I've merged with the latest dev and applied the changes requested.

andrew-platt commented:
@ebranlard suggested this be merged after #729.
